Add custom instruction_proposer support to GEPA with multimodal (dspy.Image) handling

andressrg opened the pull request:
Closes https://github.com/stanfordnlp/dspy/issues/8690

# Add support for custom instruction proposers in GEPA with multimodal handling.

As proposed by @LakshyAAAgrawal, this PR makes GEPA’s instruction-proposal step pluggable. Users can supply a custom instruction proposer, enabling GEPA to handle dspy.Image and other custom types.

## Key Features

- **Custom Instruction Proposer Protocol**: New `InstructionProposerProtocol` for implementing custom instruction generation
- **MultiModalInstructionProposer**: Built-in proposer that handles multimodal inputs (text + images) during optimization
- **Structured Image Handling**: Preserves `dspy.Image` objects without serialization in reflection LM calls
- **Backward Compatibility**: Existing GEPA usage continues to work unchanged

## Usage

### For multi-modal use cases

```python
import dspy
from dspy.teleprompt.gepa.instruction_proposal import MultiModalInstructionProposer

# Configure your language models
lm = dspy.LM(model='gpt-4o-mini')
reflection_lm = dspy.LM(model='gpt-4o', temperature=1.0, max_tokens=32000)

# Create a program that uses images
program = dspy.Predict("image: dspy.Image -> label: str")

# Set up GEPA with custom instruction proposer
gepa = dspy.GEPA(
    metric=your_metric_function,
    max_metric_calls=50,
    reflection_lm=reflection_lm,
    instruction_proposer=MultiModalInstructionProposer()  # Enable multimodal support
)

# Compile with image data
compiled_program = gepa.compile(program, trainset=image_trainset, valset=image_valset)
```

### Advanced: Custom Instruction Proposer

You can implement your own instruction proposer by following the protocol:

```python
from dspy.teleprompt.gepa.instruction_proposal import InstructionProposerProtocol

class MyCustomProposer:
    def __call__(self, current_instruction: str, reflective_dataset: list[dict]) -> str:
        # Your custom logic here
        # reflective_dataset contains examples with inputs, outputs, and feedback
        # May include dspy.Image objects in inputs
        
        # Analyze the feedback and generate improved instruction
        improved_instruction = analyze_and_improve(current_instruction, reflective_dataset)
        return improved_instruction

# Use with GEPA
gepa = dspy.GEPA(
    metric=your_metric,
    max_metric_calls=50,
    reflection_lm=reflection_lm,
    instruction_proposer=MyCustomProposer()
)
```

## Backward Compatibility

This change is fully backward compatible. Existing GEPA code will continue to work without modification:

```python
# This still works exactly as before
gepa = dspy.GEPA(metric=my_metric, max_metric_calls=50, reflection_lm=reflection_lm)
compiled = gepa.compile(program, trainset, valset)
```

## Files Changed

- `dspy/teleprompt/gepa/gepa.py`: Add `instruction_proposer` parameter
- `dspy/teleprompt/gepa/gepa_utils.py`: Integrate custom proposer logic and preserve image objects
- `dspy/teleprompt/gepa/instruction_proposal.py`: New multimodal instruction proposer implementation
- `tests/teleprompt/test_gepa_instruction_proposer.py`

This enables GEPA optimization to work with visual content while maintaining proper image handling throughout the reflection and instruction proposal process.

LakshyAAAgrawal commented:
Dear @andressrg,

Thank you so much for kindly contributing the custom instruction proposal to GEPA. The implementation is great. I have added a few suggestions. Overall I think this is good to merge.

jzabroski commented on dspy/teleprompt/gepa/gepa.py:
```diff
@@ -282,6 +283,8 @@ def __init__(
         track_best_outputs: bool = False,
         # Reproducibility
         seed: int | None = 0,
+        # Custom instruction proposer
```
`# Reflection based configuration` should also probably say `# Reflection configuration` so that the docstring and code are the same and the code comment functions as a backreference.

andressrg replied:
For sure!

Updated

andressrg commented on dspy/teleprompt/gepa/gepa_utils.py:
```diff
@@ -1,13 +1,18 @@
 from dspy.evaluate import Evaluate
 from dspy.primitives import Example, Prediction
+
+if TYPE_CHECKING:
```
correct. Removed.

andressrg commented on dspy/teleprompt/gepa/instruction_proposal.py:
```diff
@@ -0,0 +1,280 @@
+        ...
+
+
+class GenerateEnhancedMultimodalInstructionFromFeedback(dspy.Signature):
```
Glad you like them 🚀 🚀 🚀

andressrg commented on dspy/teleprompt/gepa/gepa_utils.py:
```diff
@@ -227,29 +268,3 @@ def make_reflective_dataset(self, candidate, eval_batch, components_to_update):
 
         return ret_d
-
-    # TODO: The current DSPyAdapter implementation uses the GEPA default propose_new_texts.
```
good point. Added back.

andressrg commented:
@LakshyAAAgrawal Thank you for the detailed feedback! All comments have been addressed:

- Refactored to use GEPA's native ProposalFn protocol instead of custom InstructionProposerProtocol
- Moved instruction_proposer to Reflection configuration section and updated docstring
- Added assertion for reflection_lm when using custom proposer
- Kept the TODO comment for documentation purposes

The new ProposalFn-based API is indeed much more powerful and extensible. It opens the door for multi-component instruction updates in the future as you mentioned. Great suggestion.

LakshyAAAgrawal commented on dspy/teleprompt/gepa/gepa.py:
```diff
@@ -211,6 +211,13 @@ def metric(
             for optimal performance.
         skip_perfect_score: Whether to skip examples with perfect scores during reflection. Default is True.
+        instruction_proposer: Optional custom instruction proposer implementing GEPA's ProposalFn protocol. 
+            If provided, GEPA will use this custom proposer instead of its default instruction proposal 
```
Can you include some link to the `default instruction proposal` so that people wanting to implement their own logic have a reference point? Also some kind of link to `MultiModalInstructionProposer() from dspy.teleprompt.gepa.instruction_proposal`.

andressrg replied:
I'm not sure how to add links here. I've just added this:

```
Use `MultiModalInstructionProposer()` from `dspy.teleprompt.gepa.instruction_proposal` 
for handling visual content. If None (default), GEPA uses its built-in text-optimized 
proposer (see `gepa.strategies.instruction_proposal.InstructionProposalSignature` 
for reference implementation).
```

Is this what you were proposing? or is there a way to add hrefs here that I might be familiar with? 🤔

LakshyAAAgrawal replied:
I think this is markdown so \[]() should work. I think you can feel free to ignore this as nit for now, otherwise, just include a markdown link to the source location. Won't hold back merge for this!

LakshyAAAgrawal reviewed (changes requested):
Left a few more suggestions, I think this is the final round!

andressrg commented on dspy/teleprompt/gepa/gepa_utils.py:
```diff
@@ -77,29 +75,22 @@ def __init__(
-                        )
-
-                return new_texts
+                with dspy.context(lm=self.reflection_lm):
```
I like the idea!

Just made this change:

```
        if self.custom_instruction_proposer is not None:
            # We are only overriding the propose_new_texts method when a custom
            # instruction proposer is provided. Otherwise, we use the GEPA
            # default propose_new_texts.

            def custom_propose_new_texts(
                candidate: dict[str, str],
                reflective_dataset: dict[str, list[dict[str, Any]]],
                components_to_update: list[str]
            ) -> dict[str, str]:
                if self.reflection_lm is not None:
                    with dspy.context(lm=self.reflection_lm):
                        return self.custom_instruction_proposer(
                            candidate=candidate,
                            reflective_dataset=reflective_dataset,
                            components_to_update=components_to_update
                        )
                else:
                    return self.custom_instruction_proposer(
                        candidate=candidate,
                        reflective_dataset=reflective_dataset,
                        components_to_update=components_to_update
                    )

            self.propose_new_texts = custom_propose_new_texts
```

Now, the core `gepa/appi.py` does require the user to provide a reflection_lm:

https://github.com/gepa-ai/gepa/blob/2cf10c79125533af345ffbd48497005dd1f68ee9/src/gepa/api.py#L127

Which means that in order to have true flexibility, the core gepa library needs to be updated as well I believe.

LakshyAAAgrawal replied:
Can you take a look at: https://github.com/gepa-ai/gepa/pull/50

I think this should fix it. Can you check?

andressrg replied:
Nice!. Checking now

andressrg commented:
Updated!

Left a one question, and one comment, to make sure I understand your feedback correctly.

Thanks

LakshyAAAgrawal commented on dspy/teleprompt/gepa/gepa.py:
```diff
@@ -309,11 +317,11 @@ def __init__(
         self.reflection_minibatch_size = reflection_minibatch_size
         self.candidate_selection_strategy = candidate_selection_strategy
-        # self.reflection_lm = reflection_lm
         assert reflection_lm is not None, "GEPA requires a reflection language model to be provided. Typically, you can use `dspy.LM(model='gpt-5', temperature=1.0, max_tokens=32000)` to get a good reflection model. Reflection LM is used by GEPA to reflect on the behavior of the program and propose new instructions, and will benefit from a strong model."
```
This should be modified to say "assert custom_proposer is not None or reflection_lm is not None"

andressrg replied:
sure. Done

andressrg commented:
Great! 

On it.

andressrg commented on uv.lock:
```diff
@@ -1,5 +1,5 @@
 version = 1
-revision = 2
+revision = 3
```
@LakshyAAAgrawal I'm aware my uv (version 0.8.14) updated the revision here.

Let me know if you prefer me to revert it.

LakshyAAAgrawal commented on uv.lock:
```diff
@@ -664,7 +664,7 @@ wheels = [
 [[package]]
 name = "dspy"
-version = "3.0.0"
+version = "3.0.1"
```
Probably it would be good to merge back/rebase from main and then merge this after.

andressrg commented on dspy/teleprompt/gepa/gepa.py:
```diff
@@ -210,6 +209,14 @@ def metric(
             a strong reflection model. Consider using `dspy.LM(model='gpt-5', temperature=1.0, max_tokens=32000)` 
             for optimal performance.
         skip_perfect_score: Whether to skip examples with perfect scores during reflection. Default is True.
+        instruction_proposer: Optional custom instruction proposer implementing GEPA's ProposalFn protocol. 
```
sure. On it

andressrg replied:
Added

andressrg commented:
Thank you very for the guidance and the feedback!

GEPA is amazing and I'm looking forward to all the vision use-cases that can be optimized by leveraging it!
